framework/src/utils/KDTree.C

```diff
@@ -0,0 +1,68 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "KDTree.h"
+#include "MooseError.h"
+
+#include "libmesh/nanoflann.hpp"
+
+KDTree::KDTree(std::vector<Point> & master_points)
+  : _kd_tree(nullptr), _point_list_adaptor(master_points)
+{
+}
+
+void
+KDTree::buildTree()
+{
+  // Maximum number of points in each leaf of the Kd tree
+  unsigned int max_leaf_size = 10;
```

Comment:
The Kdtree basically divides the master point cloud into bins based on the distance. This parameter specifies the maximum number of points that can be present in a bin. As the max_leaf_size increases, the time taken to  build the tree decreases but the time taken to conduct the search increases. I don't know if there is an optimal value that works well for all problems.

I took this value from the examples within libmesh/contrib/nanoflann. I could add this as another parameter (like patch size) in MooseMesh with default value of 10 and users can change it if needed. Would that be okay?